packages/lib-sourcify/src/Verification/Transformations.ts

```diff
@@ -0,0 +1,348 @@
+  libraries?: {
+    [id: string]: string;
+  };
+  immutables?: {
+    [id: string]: string;
+  };
+  cborAuxdata?: {
+    [id: string]: string;
+  };
+}
+
+// returns the full bytecode with the call protection replaced with the real address
+export function extractCallProtectionTransformation(
+  normalizedRecompiledBytecode: string,
+  onchainRuntimeBytecode: string,
+) {
+  const transformations: Transformation[] = [];
+  const transformationValues: TransformationValues = {};
+  const template = normalizedRecompiledBytecode;
+  const real = onchainRuntimeBytecode;
+
+  const push20CodeOp = '73';
+  const callProtection = `0x${push20CodeOp}${'00'.repeat(20)}`;
+
+  if (template.startsWith(callProtection)) {
+    const replacedCallProtection = real.slice(0, 0 + callProtection.length);
+    const callProtectionAddress = replacedCallProtection.slice(4); // remove 0x73
+    transformations.push(CallProtectionTransformation());
+    transformationValues.callProtection = '0x' + callProtectionAddress;
+
+    return {
+      normalizedRecompiledBytecode:
+        replacedCallProtection + template.substring(callProtection.length),
+      transformations,
+      transformationValues,
+    };
+  }
+  return {
+    normalizedRecompiledBytecode: template,
+    transformations,
+    transformationValues,
+  };
+}
+
+/**
+ * Replaces the values of the immutable variables in the (onchain) deployed bytecode with zeros, so that the bytecode can be compared with the (offchain) recompiled bytecode.
+ * Easier this way because we can simply replace with zeros
+ * Example immutableReferences: {"97":[{"length":32,"start":137}],"99":[{"length":32,"start":421}]} where 97 and 99 are the AST ids
+ */
+export function extractImmutablesTransformation(
+  normalizedRecompiledBytecodeWith0x: string,
+  onchainRuntimeBytecodeWith0x: string,
+  immutableReferences: ImmutableReferences,
+  auxdataStyle: AuxdataStyle,
+) {
+  const transformations: Transformation[] = [];
+  const transformationValues: TransformationValues = {};
+  // Remove "0x" from the beginning of both bytecodes.
+  const onchainRuntimeBytecode = onchainRuntimeBytecodeWith0x.slice(2);
+  let normalizedRecompiledBytecode =
+    normalizedRecompiledBytecodeWith0x.slice(2);
+
+  Object.keys(immutableReferences).forEach((astId) => {
+    immutableReferences[astId].forEach((reference) => {
+      const { start, length } = reference;
+
+      // Save the transformation
+      transformations.push(
+        ImmutablesTransformation(
+          start,
+          astId,
+          auxdataStyle === AuxdataStyle.SOLIDITY ? 'replace' : 'insert',
+        ),
+      );
+
+      // Extract the immutable value from the onchain bytecode.
+      const immutableValue = onchainRuntimeBytecode.slice(
+        start * 2,
+        start * 2 + length * 2,
+      );
+
+      // Save the transformation value
+      if (transformationValues.immutables === undefined) {
+        transformationValues.immutables = {};
+      }
```

Comment:
What's the purpose of this check? Isn't it always `undefined`? we assign `{}` to `transformationValues` just a few lines above.

Reply:
We assign `const transformationValues: TransformationValues = {};` but not

```ts
const transformationValues: TransformationValues = {
  immutables: {}
};
```

then we enter a loop in which multiple immutables are stored. For the first one `transformationValues.immutables` is empty and we have to initialize it.

Reply:
Ok I see. Just, this check always passes so we can just assign {} inside the loop without the if. 

Just a small thing

Reply:
You mean outside the loop at the beginning? If we were to assign it without the if we would override any previously assigned key/value

Reply:
Ahh ok now I see. This check does not always pass, it will only pass in the first loop run. 

Just ignore my comment.

Reply:
np :)